lib/ipam/ipam.go

```diff
@@ -103,27 +103,162 @@ func (c ipamClient) AutoAssign(ctx context.Context, args AutoAssignArgs) ([]net.
+func (c ipamClient) getBlockFromAffinity(ctx context.Context, aff *model.KVPair) (*model.KVPair, error) {
+	// Parse out affinity data.
+	cidr := aff.Key.(model.BlockAffinityKey).CIDR
+	host := aff.Key.(model.BlockAffinityKey).Host
+	state := aff.Value.(*model.BlockAffinity).State
+	logCtx := log.WithFields(log.Fields{"host": host, "cidr": cidr})
+
+	// Get the block referenced by this affinity.
+	logCtx.Info("Attempting to load block")
+	b, err := c.client.Get(ctx, model.BlockKey{CIDR: cidr}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+			// The block referenced by the affinity doesn't exist. Try to create it.
+			logCtx.Info("The referenced block doesn't exist, trying to create it")
+			aff.Value.(*model.BlockAffinity).State = model.StatePending
+			aff, err = c.client.Update(ctx, aff)
+			if err != nil {
+				logCtx.WithError(err).Warn("Error updating block affinity")
+				return nil, err
+			}
+			logCtx.Info("Wrote affinity as pending")
+
+			cfg, err := c.GetIPAMConfig(ctx)
+			if err != nil {
+				logCtx.WithError(err).Errorf("Error getting IPAM Config")
+				return nil, err
+			}
+
+			// Claim the block, which will also confirm the affinity.
+			logCtx.Info("Attempting to claim the block")
+			b, err := c.blockReaderWriter.claimAffineBlock(ctx, aff, *cfg)
+			if err != nil {
+				logCtx.WithError(err).Warn("Error claiming block")
+				return nil, err
+			}
+			return b, nil
+		}
+		logCtx.WithError(err).Error("Error getting block")
+		return nil, err
+	}
+
+	// If the block doesn't match the affinity, it means we've got a stale affininty hanging around.
+	// We should remove it.
+	blockAffinity := b.Value.(*model.AllocationBlock).Affinity
+	if blockAffinity == nil || *blockAffinity != fmt.Sprintf("host:%s", host) {
+		logCtx.WithField("blockAffinity", blockAffinity).Warn("Block does not match the provided affinity, deleting stale affinity")
+		_, err := c.client.Delete(ctx, aff.Key, aff.Revision)
+		if err != nil {
+			logCtx.WithError(err).Warn("Error deleting stale affinity")
+			return nil, err
+		}
+		return nil, errStaleAffinity(fmt.Sprintf("Affinity is stale: %+v", aff))
+	}
+
+	// If the block does match the affinity but the affinity has not been confirmed,
+	// try to confirm it. Treat empty string as confirmed for compatibility with older data.
+	if state != model.StateConfirmed && state != "" {
+		// Write the affinity as pending.
+		logCtx.Info("Affinity has not been confirmed - attempt to confirm it")
+		aff.Value.(*model.BlockAffinity).State = model.StatePending
+		aff, err = c.client.Update(ctx, aff)
+		if err != nil {
+			logCtx.WithError(err).Warn("Error marking affinity as pending as part of confirmation process")
+			return nil, err
+		}
+
+		// CAS the block to get a new revision and invalidate any other instances
+		// that might be trying to operate on the block.
+		logCtx.Info("Writing block to get a new revision")
+		b, err = c.client.Update(ctx, b)
```

Comment:
Is this used for preventing other process on the same node operating this block?  Why is there any other process? Like what?

Reply:
@mdianjun yeah, there might be other clients trying to perform an IPAM allocation at the same time, e.g. multiple pods starting on the host at the same time. We need to make sure that only one client operates on the block at a time. If two clients conflict, then one should succeed and the other should retry.